Feature: Added experimental terminal integration

gave92 opened the pull request:
**Resolved / Related Issues**
- [x] Were these changes approved in an issue or discussion with the project maintainers? In order to prevent extra work, feature requests and changes to the codebase must be approved before the pull request will be reviewed. This prevents extra work for the contributors and maintainers.
   Closes #6235

**Validation**
How did you test these changes?
- [x] Did you build the app and test your changes?

**Screenshots (optional)**
<img width="960" alt="image" src="https://github.com/files-community/Files/assets/9673091/e1d446e2-ac6b-475f-9675-7c407d0abee5">

gave92 committed:
[WIP]

gave92 committed:
[WIP]

gave92 committed:
[WIP]

gave92 committed:
Initial work

gave92 committed:
Add build files

gave92 committed:
Added credits to source files

gave92 committed:
Add build files for xterm.js

gave92 committed:
Implement copy and resize

gave92 committed:
Set background and hide double scrollbars

gave92 committed:
Toggle button & cleanup

gave92 committed:
Start from current folder

gave92 committed:
Sync folder to cmd

gave92 committed:
Implemented up sync button

gave92 committed:
Enable profile selection

gave92 committed:
Updated comments

gave92 committed:
Test theme change

gave92 committed:
Place classes in project structure

gave92 committed:
Added missing attribution

gave92 committed:
Fix webview scrollbars in dark mode

gave92 committed:
Moved Terminal under Utils

gave92 committed:
Removed extra colon

gave92 committed:
Fix Release build

gave92 committed:
Set background color in html

gave92 committed:
Fix buttons for wsl

gave92 committed:
Fix webview flashing white

hez2010 commented:
I'm wondering if we can use the conhost from Windows Terminal instead of using the WebView-based terminal?

gave92 commented:
I'd like to but it's huge and I could not find an easy way to integrate it.

0x5bfa commented:
What's left to do?
Looks solid!

gave92 committed:
Remove extra commands

gave92 committed:
Improve scrollbar

gave92 commented:
Hey @0x5bfa the TODOs I have in mind are:
- Confirm that the chosen approach (web view + xterm) is ok
- Define different terminal themes for light/dark mode (if needed)
- Add settings to enable/disable the feature and to store the preferred terminal.
- Some refactoring for TerminalModels.cs (e.g. it contains more that one class)

Unfortunately these days I have very little coding time, so any help completing this is much appreciated 😊

gave92 committed:
Use windows style scrollbars

gumbarros commented on src/Files.App/Extensions/WebView2Extensions.cs:
```diff
@@ -0,0 +1,222 @@
+			sb.AppendLine($"window.{name} = {{ ");
+
+			// Test webview for our sanity
+			await webview.ExecuteScriptAsync($@"console.log(""Sanity check from iMessage"");");
```
Is this needed on production?

gumbarros commented on src/Files.App/Utils/Terminal/ConPTY/Native/ConsoleApi.cs:
```diff
@@ -0,0 +1,44 @@
+		internal const uint ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x0004;
+		internal const uint DISABLE_NEWLINE_AUTO_RETURN = 0x0008;
+
+		[DllImport("kernel32.dll", EntryPoint = "AllocConsole", SetLastError = true, CharSet = CharSet.Auto, CallingConvention = CallingConvention.StdCall)]
+		internal static extern bool AllocConsole();
+
+		[DllImport("kernel32.dll", SetLastError = true)]
+		internal static extern nint GetConsoleWindow();
+
+		[DllImport("kernel32.dll", SetLastError = true)]
+		internal static extern SafeFileHandle GetStdHandle(int nStdHandle);
+
+		[DllImport("kernel32.dll", SetLastError = true)]
+		internal static extern bool SetConsoleMode(SafeFileHandle hConsoleHandle, uint mode);
+
+		[DllImport("kernel32.dll", SetLastError = true)]
+		internal static extern bool GetConsoleMode(SafeFileHandle handle, out uint mode);
```
You can use CsWin32 for this guys, but not deal breaking in this PR I think.

gave92 replied:
Switched to CsWin32 for most methods 👍

gumbarros commented on src/Files.App/Extensions/WebView2Extensions.cs:
```diff
@@ -0,0 +1,222 @@
+			public string Value { get; set; }
+		}
+
+		public static List<TypedEventHandler<WebView2, CoreWebView2WebMessageReceivedEventArgs>> _handlers = new List<TypedEventHandler<WebView2, CoreWebView2WebMessageReceivedEventArgs>>();
```
You probably can use a [type alias](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-12.0/using-alias-types) for this `List<TypedEventHandler<WebView2, CoreWebView2WebMessageReceivedEventArgs>>` big guy.

gumbarros reviewed (commented):
Congrats for the PR, looking forward to use this feature

gave92 committed:
Code suggestions

gave92 committed:
Switch native methods to CsWin32

0x5bfa commented:
@gave92 Thank you for using CsWin32! This is our short term goal, which is to use source generator everywhere.

Commit:
Added settings under experimental

gave92 commented:
I've added a setting under Experimental to enable/disable the feature (default: off). The feature may be good enough for first release :)

yaira2 commented:
I had some notes on this but I'll have another look since it's been a while.

yaira2 commented:
The main point to figure out before merging is whether the terminal is global, or if there can be a separate terminal for each tab & pane. I lean towards the second option (similar to VS Code).

Commit:
WIP

gave92 commented:
Holding off while @hez2010 checks if it's possible to use Windows Terminal control.

gave92 committed:
Call AddWebAllowedObject after navigation

gave92 committed:
Multiple terminals

gave92 committed:
Hide pane when no terminals

gave92 commented:
Added ability to have more than one terminal 😇

yaira2 commented:
Switching between terminals is working nicely!

gave92 committed:
Switch to CsWin32 for process api

gave92 committed:
Add terminal model

gave92 committed:
Minor change

gave92 committed:
Minor change

XTorLukas commented on src/Files.App/UserControls/StatusBar.xaml:
```diff
@@ -393,5 +395,141 @@
+												Click="TerminalCloseButton_Click"
+												Tag="{x:Bind Id}"
+												ToolTipService.ToolTip="{helpers:ResourceString Name=Close}">
+												<FontIcon FontSize="12" Glyph="&#xE74D;" />
```
```suggestion
												<FontIcon FontSize="12" Glyph="&#xE711;" />
```

Use close symbol 
![image](https://github.com/files-community/Files/assets/84145589/2654e518-aa53-40ce-9633-60dc1447ef1a)

XTorLukas commented on src/Files.App/UserControls/StatusBar.xaml:
```diff
@@ -393,5 +395,141 @@
+				x:Load="{x:Bind MainPageViewModel.IsTerminalViewOpen, Mode=OneWay}"
+				Background="Transparent"
+				BorderBrush="Transparent"
+				Command="{x:Bind MainPageViewModel.TerminalSyncDownCommand}">
```
```suggestion
				Command="{x:Bind MainPageViewModel.TerminalSyncDownCommand}"
				ToolTipService.ToolTip="{helpers:ResourceString Name=SyncFolderDown}">
```
Add localized tooltip

XTorLukas commented on src/Files.App/UserControls/StatusBar.xaml:
```diff
@@ -393,5 +395,141 @@
+				x:Load="{x:Bind MainPageViewModel.IsTerminalViewOpen, Mode=OneWay}"
+				Background="Transparent"
+				BorderBrush="Transparent"
+				Command="{x:Bind MainPageViewModel.TerminalSyncUpCommand}">
```
```suggestion
				Command="{x:Bind MainPageViewModel.TerminalSyncUpCommand}"
				ToolTipService.ToolTip="{helpers:ResourceString Name=SyncFolderUp}">
```
Add localized tooltip

yaira2 replied:
@XTorLukas there is a good chance these buttons won't end up being included. At this point we're still working on the best way to embed a Terminal into the window.

XTorLukas commented:
It looks good 😊 I would just add localized text and tooltip to the buttons

XTorLukas reviewed (changes requested):
I would just add localized text and tooltip to the buttons 😊

XTorLukas commented on src/Files.App/Strings/en-US/Resources.resw:
```diff
@@ -3838,4 +3838,7 @@
   </data>
+  <data name="SettingsTerminalIntegration" xml:space="preserve">
+    <value>Enable Terminal integration</value>
+  </data>
```
But the name for `SyncFolderDown` --> `Down?` I don't know how to specify
The same goes for `SyncFolderUp` --> `Up?`
But it will be localized

yair100 committed:
Added margin

0x5bfa commented on src/Files.App/Utils/Terminal/UI/Build/src/style.css:
```diff
@@ -0,0 +1,29 @@
+    -ms-content-zooming: none;
+}
+
+#terminal-container {
```
@yaira2 you might want to try here.

```css
border-radius: 8px;
```

gave92 committed:
Readded Newtonsoft.Json dep

gave92 commented:
Rounded corners 💪

![image](https://github.com/files-community/Files/assets/9673091/d6fb1182-7e5a-40ff-aeec-b2a28fc94351)

0x5bfa commented:
Damn, over 9000 addition. Would it be better to add round corner into css file?

0x5bfa commented:
Omg, over 9000 addition. Would it be better to add round corner into css file?

yaira2 commented on src/Files.App/Files.App.csproj:
```diff
@@ -79,6 +79,7 @@
         <PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0" />
         <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.0" />
         <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.3233" />
+        <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
But why 😂

gave92 replied:
Can't get WebViewExtsnsions to work with System.Text :(

yaira2 replied:
I tried converting the code to `System.Text.Json` but ran into an issue where the terminal text wouldn't display.

gave92 replied:
Done!

yaira2 commented:
> Omg, over 9000 addition. Would it be better to add round corner into css file?

If possible, but we haven't found a way to do this.

gave92 commented:
Lol yeah completely new projects just for having rounded corners xD I do hope MS will add this to standard webview.

gave92 committed:
Stupid solution n.1

yaira2 commented:
I won't mention it for now 😂

gave92 committed:
Switched WebView2Extensions to System.Text.Json

yair100 committed:
Switch status bar with terminal

Revert "Switch status bar with terminal"

This reverts commit 8d793ed305bdb71cd7eb144c85457da274540bc4.

Update MainPage.xaml

Revert "Switch status bar with terminal"

This reverts commit 248347a5e554a9dfd140dbeb969fbe1079a29684.

Update StatusBar.xaml

Lamparter reviewed (approved):
Apart from the merge conflicts, what's stopping this from being merged? Was there something else to implement or is it not stable enough?

yaira2 commented:
Functionality wise this is pretty much good to go, but we haven't settled on a design yet.

gave92 committed:
Remove duplicates from NativeMethods.txt

Commit:
Remove duplicate SetWindowLongPtr

Commit:
Fix build errors

Commit:
Restore functionalities

Commit:
Restore sync path functionality

yair100 committed:
Refactor WebView2 binding and terminal fixes

yair100 committed:
Updated
